File: plugins/vscode/src/refactBinaryResolver.ts
import { spawn } from "child_process";
import * as crypto from "crypto";
import * as fs from "fs";
import * as http from "http";
import * as https from "https";
import * as os from "os";
import * as path from "path";
import { compareVersions } from "./refactDaemon";

export const REFACT_RELEASE_BASE_URL = "https://github.com/JegernOUTT/refact/releases/download";
const USER_AGENT_HEADER = "User-Agent";

export type RefactReleaseAsset = {
    target: string;
    archiveName: string;
    archiveUrl: string;
    sha256Url: string;
};

export type RefactBinaryResolverOptions = {
    explicitPath?: string;
    minVersion: string;
    pinnedVersion: string;
    cacheDir: string;
    pathEnv?: string;
    homeDir?: string;
    platform?: string;
    arch?: string;
    runVersion?: (binPath: string) => Promise<string | undefined>;
    downloadFile?: (url: string, destPath: string) => Promise<void>;
    extractArchive?: (archivePath: string, destDir: string, platform: string) => Promise<void>;
    chmod?: (binPath: string) => Promise<void>;
};

export function binaryNameForPlatform(platform: string = process.platform): string {
    return platform === "win32" ? "refact.exe" : "refact";
}

export function refactReleaseTarget(platform: string = process.platform, arch: string = process.arch): string {
    if (platform === "win32") {
        if (arch === "x64") { return "x86_64-pc-windows-msvc"; }
        if (arch === "ia32") { return "i686-pc-windows-msvc"; }
        if (arch === "arm64") { return "aarch64-pc-windows-msvc"; }
    }
    if (platform === "linux") {
        if (arch === "x64") { return "x86_64-unknown-linux-gnu"; }
        if (arch === "arm64") { return "aarch64-unknown-linux-gnu"; }
    }
    if (platform === "darwin") {
        if (arch === "x64") { return "x86_64-apple-darwin"; }
        if (arch === "arm64") { return "aarch64-apple-darwin"; }
    }
    throw new Error(`unsupported Refact release target for ${platform}/${arch}`);
}

export function refactReleaseAsset(version: string, target: string, platform: string = process.platform): RefactReleaseAsset {
    const extension = platform === "win32" ? "zip" : "tar.gz";
    const archiveName = `refact-${version}-${target}.${extension}`;
    const archiveUrl = `${REFACT_RELEASE_BASE_URL}/engine/v${version}/${archiveName}`;
    return {
        target,
        archiveName,
        archiveUrl,
        sha256Url: `${archiveUrl}.sha256`,
    };
}

export function extractRefactVersion(output: string | undefined): string | undefined {
    const text = output?.trim();
    if (!text) { return undefined; }
    const refactMatch = text.match(/(?:^|\s)refact\s+([0-9]+(?:\.[0-9]+){0,2}(?:[-+][0-9A-Za-z._-]+)?)/i);
    if (refactMatch?.[1]) { return refactMatch[1]; }
    const genericMatch = text.match(/([0-9]+(?:\.[0-9]+){1,2}(?:[-+][0-9A-Za-z._-]+)?)/);
    return genericMatch?.[1];
}

export async function resolveRefactBinary(options: RefactBinaryResolverOptions): Promise<string> {
    const explicitPath = options.explicitPath?.trim();
    if (explicitPath) {
        return path.resolve(explicitPath);
    }

    const minVersion = options.minVersion;
    const platform = options.platform ?? process.platform;
    const arch = options.arch ?? process.arch;
    const runVersion = options.runVersion ?? readRefactVersion;
    for (const candidate of systemRefactCandidates(options.pathEnv ?? process.env.PATH ?? "", options.homeDir ?? os.homedir(), platform)) {
        if (await isCompatibleRefactBinary(candidate, minVersion, runVersion)) {
            return candidate;
        }
    }

    return downloadPinnedRefactBinary({ ...options, platform, arch, runVersion });
}

function systemRefactCandidates(pathEnv: string, homeDir: string, platform: string): string[] {
    const binaryName = binaryNameForPlatform(platform);
    const candidates = pathEnv
        .split(path.delimiter)
        .filter(entry => entry.trim().length > 0)
        .map(entry => path.join(entry, binaryName));
    candidates.push(path.join(homeDir, ".refact", "bin", binaryName));
    return Array.from(new Set(candidates.map(candidate => path.resolve(candidate))));
}

async function isCompatibleRefactBinary(
    binPath: string,
    minVersion: string,
    runVersion: (binPath: string) => Promise<string | undefined>,
): Promise<boolean> {
    if (!fileExists(binPath)) { return false; }
    const version = extractRefactVersion(await runVersion(binPath));
    return !!version && compareVersions(version, minVersion) >= 0;
}

async function downloadPinnedRefactBinary(options: Required<Pick<RefactBinaryResolverOptions, "platform" | "arch" | "runVersion">> & RefactBinaryResolverOptions): Promise<string> {
    const target = refactReleaseTarget(options.platform, options.arch);
    const binaryName = binaryNameForPlatform(options.platform);
    const targetDir = path.join(options.cacheDir, options.pinnedVersion, target);
    const binPath = path.join(targetDir, binaryName);
    if (await isCompatibleRefactBinary(binPath, options.minVersion, options.runVersion)) {
        return binPath;
    }

    const downloadFile = options.downloadFile ?? defaultDownloadFile;
    const extractArchive = options.extractArchive ?? defaultExtractArchive;
    const chmod = options.chmod ?? defaultChmod;
    const asset = refactReleaseAsset(options.pinnedVersion, target, options.platform);
    const tmpDir = path.join(options.cacheDir, `tmp-${process.pid}-${Date.now()}-${Math.random().toString(16).slice(2)}`);
    const archivePath = path.join(tmpDir, asset.archiveName);
    const shaPath = `${archivePath}.sha256`;
    const extractDir = path.join(tmpDir, "extract");
    await fs.promises.mkdir(extractDir, { recursive: true });
    try {
        await downloadFile(asset.archiveUrl, archivePath);
        await downloadFile(asset.sha256Url, shaPath);
        await verifySha256(archivePath, shaPath);
        await extractArchive(archivePath, extractDir, options.platform);
        const extractedBin = path.join(extractDir, binaryName);
        if (!fileExists(extractedBin)) {
            throw new Error(`downloaded Refact archive did not contain ${binaryName}`);
        }
        await chmod(extractedBin);
        await fs.promises.rm(targetDir, { recursive: true, force: true });
        await fs.promises.mkdir(path.dirname(targetDir), { recursive: true });
        await fs.promises.rename(extractDir, targetDir);
        await chmod(binPath);
        if (!await isCompatibleRefactBinary(binPath, options.minVersion, options.runVersion)) {
            throw new Error(`downloaded Refact binary is older than ${options.minVersion}`);
        }
        return binPath;
    } finally {
        await fs.promises.rm(tmpDir, { recursive: true, force: true }).catch(() => undefined);
    }
}

async function readRefactVersion(binPath: string): Promise<string | undefined> {
    return runAndCapture(binPath, ["--version"], 5000);
}

function fileExists(candidate: string): boolean {
    try {
        return fs.statSync(candidate).isFile();
    } catch {
        return false;
    }
}

async function verifySha256(archivePath: string, shaPath: string): Promise<void> {
    const expected = expectedSha256(await fs.promises.readFile(shaPath, "utf8"));
    const actual = await sha256File(archivePath);
    if (actual !== expected) {
        throw new Error(`sha256 mismatch for ${path.basename(archivePath)}`);
    }
}

function expectedSha256(text: string): string {
    const match = text.match(/[a-fA-F0-9]{64}/);
    if (!match) {
        throw new Error("sha256 sidecar did not contain a checksum");
    }
    return match[0].toLowerCase();
}

function sha256File(filePath: string): Promise<string> {
    return new Promise((resolve, reject) => {
        const hash = crypto.createHash("sha256");
        const stream = fs.createReadStream(filePath);
        stream.on("data", chunk => hash.update(chunk));
        stream.once("error", reject);
        stream.once("end", () => resolve(hash.digest("hex")));
    });
}

function defaultDownloadFile(url: string, destPath: string): Promise<void> {
    return new Promise((resolve, reject) => {
        downloadFileToPath(url, destPath, 0, resolve, reject);
    });
}

function downloadFileToPath(
    url: string,
    destPath: string,
    redirects: number,
    resolve: () => void,
    reject: (error: Error) => void,
): void {
    const client = url.startsWith("http:") ? http : https;
    const request = client.get(url, { headers: { [USER_AGENT_HEADER]: "refact-vscode" } }, response => {
        const statusCode = response.statusCode ?? 0;
        const location = response.headers.location;
        if (statusCode >= 300 && statusCode < 400 && location && redirects < 5) {
            response.resume();
            downloadFileToPath(new URL(location, url).toString(), destPath, redirects + 1, resolve, reject);
            return;
        }
        if (statusCode !== 200) {
            response.resume();
            reject(new Error(`download failed ${statusCode} ${url}`));
            return;
        }
        fs.mkdirSync(path.dirname(destPath), { recursive: true });
        const output = fs.createWriteStream(destPath);
        response.pipe(output);
        output.once("finish", () => output.close(resolve));
        output.once("error", error => reject(error));
    });
    request.once("error", error => reject(error));
}

async function defaultExtractArchive(archivePath: string, destDir: string, platform: string): Promise<void> {
    try {
        await runProcess("tar", archivePath.endsWith(".zip") ? ["-xf", archivePath, "-C", destDir] : ["-xzf", archivePath, "-C", destDir]);
    } catch (error) {
        if (platform !== "win32" || !archivePath.endsWith(".zip")) {
            throw error;
        }
        await runProcess("powershell.exe", [
            "-NoProfile",
            "-NonInteractive",
            "-ExecutionPolicy",
            "Bypass",
            "-Command",
            "$ErrorActionPreference='Stop'; Expand-Archive -LiteralPath $args[0] -DestinationPath $args[1] -Force",
            archivePath,
            destDir,
        ]);
    }
}

async function defaultChmod(binPath: string): Promise<void> {
    if (process.platform !== "win32") {
        await fs.promises.chmod(binPath, 0o755);
    }
}

function runAndCapture(command: string, args: string[], timeoutMs: number): Promise<string | undefined> {
    return new Promise(resolve => {
        const child = spawn(command, args, { stdio: ["ignore", "pipe", "pipe"] });
        const chunks: Buffer[] = [];
        let timedOut = false;
        const timer = setTimeout(() => {
            timedOut = true;
            child.kill();
        }, timeoutMs);
        child.stdout?.on("data", chunk => chunks.push(Buffer.from(chunk)));
        child.stderr?.on("data", chunk => chunks.push(Buffer.from(chunk)));
        child.once("error", () => {
            clearTimeout(timer);
            resolve(undefined);
        });
        child.once("close", code => {
            clearTimeout(timer);
            if (timedOut || code !== 0) {
                resolve(undefined);
                return;
            }
            resolve(Buffer.concat(chunks).toString("utf8"));
        });
    });
}

function runProcess(command: string, args: string[]): Promise<void> {
    return new Promise((resolve, reject) => {
        const child = spawn(command, args, { stdio: ["ignore", "ignore", "pipe"] });
        const stderr: Buffer[] = [];
        child.stderr?.on("data", chunk => stderr.push(Buffer.from(chunk)));
        child.once("error", reject);
        child.once("close", code => {
            if (code === 0) {
                resolve();
                return;
            }
            reject(new Error(`${command} exited with ${code}: ${Buffer.concat(stderr).toString("utf8")}`));
        });
    });
}

File: plugins/intellij/src/main/kotlin/com/smallcloud/refactai/lsp/RefactBinaryResolver.kt
package com.smallcloud.refactai.lsp

import com.intellij.openapi.util.SystemInfo
import java.io.File
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URI
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.security.MessageDigest
import java.util.Comparator
import java.util.concurrent.TimeUnit
import java.util.zip.ZipInputStream

internal const val REFACT_RELEASE_BASE_URL = "https://github.com/JegernOUTT/refact/releases/download"

internal data class RefactReleaseAsset(
    val target: String,
    val archiveName: String,
    val archiveUrl: String,
    val sha256Url: String,
)

internal data class RefactBinaryResolverOptions(
    val explicitPath: String? = null,
    val minVersion: String,
    val pinnedVersion: String,
    val cacheDir: Path,
    val pathEnv: String = System.getenv("PATH").orEmpty(),
    val homeDir: Path = Path.of(System.getProperty("user.home")),
    val osName: String = System.getProperty("os.name"),
    val arch: String = System.getProperty("os.arch"),
    val versionReader: (Path) -> String? = ::readRefactVersion,
    val downloader: (URI, Path) -> Unit = ::downloadFile,
    val extractor: (Path, Path, Boolean) -> Unit = ::extractArchive,
    val chmod: (Path) -> Unit = ::makeExecutable,
)

internal object RefactBinaryResolver {
    fun resolve(options: RefactBinaryResolverOptions): String {
        val explicit = options.explicitPath?.trim()?.takeIf { it.isNotEmpty() }
        if (explicit != null) {
            return Path.of(explicit).toAbsolutePath().normalize().toString()
        }

        for (candidate in systemRefactCandidates(options.pathEnv, options.homeDir, options.osName)) {
            if (isCompatibleRefactBinary(candidate, options.minVersion, options.versionReader)) {
                return candidate.toString()
            }
        }

        return downloadPinnedRefactBinary(options)
    }
}

internal fun refactBinaryName(osName: String = System.getProperty("os.name")): String {
    return if (osName.lowercase().contains("win")) "refact.exe" else "refact"
}

internal fun refactReleaseTarget(osName: String = System.getProperty("os.name"), arch: String = System.getProperty("os.arch")): String {
    val os = osName.lowercase()
    val normalizedArch = when (arch.lowercase()) {
        "amd64", "x86_64" -> "x86_64"
        "x86", "i386", "i686" -> "i686"
        "aarch64", "arm64" -> "aarch64"
        else -> arch.lowercase()
    }
    return when {
        os.contains("win") && normalizedArch == "x86_64" -> "x86_64-pc-windows-msvc"
        os.contains("win") && normalizedArch == "i686" -> "i686-pc-windows-msvc"
        os.contains("win") && normalizedArch == "aarch64" -> "aarch64-pc-windows-msvc"
        os.contains("linux") && normalizedArch == "x86_64" -> "x86_64-unknown-linux-gnu"
        os.contains("linux") && normalizedArch == "aarch64" -> "aarch64-unknown-linux-gnu"
        os.contains("mac") && normalizedArch == "x86_64" -> "x86_64-apple-darwin"
        os.contains("mac") && normalizedArch == "aarch64" -> "aarch64-apple-darwin"
        else -> throw IllegalArgumentException("unsupported Refact release target for $osName/$arch")
    }
}

internal fun refactReleaseAsset(version: String, target: String, osName: String = System.getProperty("os.name")): RefactReleaseAsset {
    val extension = if (osName.lowercase().contains("win")) "zip" else "tar.gz"
    val archiveName = "refact-$version-$target.$extension"
    val archiveUrl = "$REFACT_RELEASE_BASE_URL/engine/v$version/$archiveName"
    return RefactReleaseAsset(
        target = target,
        archiveName = archiveName,
        archiveUrl = archiveUrl,
        sha256Url = "$archiveUrl.sha256",
    )
}

internal fun extractRefactVersion(output: String?): String? {
    val text = output?.trim()?.takeIf { it.isNotEmpty() } ?: return null
    Regex("""(?:^|\s)refact\s+([0-9]+(?:\.[0-9]+){0,2}(?:[-+][0-9A-Za-z._-]+)?)""", RegexOption.IGNORE_CASE)
        .find(text)
        ?.groupValues
        ?.getOrNull(1)
        ?.let { return it }
    return Regex("""([0-9]+(?:\.[0-9]+){1,2}(?:[-+][0-9A-Za-z._-]+)?)""")
        .find(text)
        ?.groupValues
        ?.getOrNull(1)
}

internal fun compareRefactVersions(left: String?, right: String?): Int {
    val leftParts = parseVersion(left)
    val rightParts = parseVersion(right)
    for (index in 0..2) {
        val diff = leftParts[index] - rightParts[index]
        if (diff != 0) return if (diff > 0) 1 else -1
    }
    return 0
}

private fun parseVersion(version: String?): List<Int> {
    val parts = version.orEmpty()
        .trim()
        .split(Regex("[.\\-+_\\s]"))
        .map { part -> Regex("^\\d+").find(part)?.value?.toIntOrNull() ?: 0 }
    return listOf(
        parts.getOrElse(0) { 0 },
        parts.getOrElse(1) { 0 },
        parts.getOrElse(2) { 0 },
    )
}

private fun systemRefactCandidates(pathEnv: String, homeDir: Path, osName: String): List<Path> {
    val binaryName = refactBinaryName(osName)
    val candidates = pathEnv.split(File.pathSeparator)
        .asSequence()
        .filter { it.isNotBlank() }
        .map { Path.of(it, binaryName).toAbsolutePath().normalize() }
        .toMutableList()
    candidates.add(homeDir.resolve(".refact").resolve("bin").resolve(binaryName).toAbsolutePath().normalize())
    return candidates.distinctBy { it.toString() }
}

private fun isCompatibleRefactBinary(binPath: Path, minVersion: String, versionReader: (Path) -> String?): Boolean {
    if (!Files.isRegularFile(binPath)) return false
    val version = extractRefactVersion(versionReader(binPath)) ?: return false
    return compareRefactVersions(version, minVersion) >= 0
}

private fun downloadPinnedRefactBinary(options: RefactBinaryResolverOptions): String {
    val target = refactReleaseTarget(options.osName, options.arch)
    val binaryName = refactBinaryName(options.osName)
    val targetDir = options.cacheDir.resolve(options.pinnedVersion).resolve(target)
    val binPath = targetDir.resolve(binaryName)
    if (isCompatibleRefactBinary(binPath, options.minVersion, options.versionReader)) {
        return binPath.toString()
    }

    val asset = refactReleaseAsset(options.pinnedVersion, target, options.osName)
    val tmpDir = options.cacheDir.resolve("tmp-${ProcessHandle.current().pid()}-${System.nanoTime()}")
    val archivePath = tmpDir.resolve(asset.archiveName)
    val shaPath = tmpDir.resolve("${asset.archiveName}.sha256")
    val extractDir = tmpDir.resolve("extract")
    Files.createDirectories(extractDir)
    try {
        options.downloader(URI(asset.archiveUrl), archivePath)
        options.downloader(URI(asset.sha256Url), shaPath)
        verifySha256(archivePath, shaPath)
        options.extractor(archivePath, extractDir, options.osName.lowercase().contains("win"))
        val extractedBin = extractDir.resolve(binaryName)
        if (!Files.isRegularFile(extractedBin)) {
            throw IOException("downloaded Refact archive did not contain $binaryName")
        }
        options.chmod(extractedBin)
        deleteRecursively(targetDir)
        Files.createDirectories(targetDir.parent)
        Files.move(extractDir, targetDir, StandardCopyOption.REPLACE_EXISTING)
        options.chmod(binPath)
        if (!isCompatibleRefactBinary(binPath, options.minVersion, options.versionReader)) {
            throw IOException("downloaded Refact binary is older than ${options.minVersion}")
        }
        return binPath.toString()
    } finally {
        deleteRecursively(tmpDir)
    }
}

private fun readRefactVersion(binPath: Path): String? {
    return try {
        val process = ProcessBuilder(binPath.toString(), "--version")
            .redirectErrorStream(true)
            .start()
        if (!process.waitFor(5, TimeUnit.SECONDS)) {
            process.destroyForcibly()
            return null
        }
        if (process.exitValue() != 0) return null
        process.inputStream.bufferedReader().readText()
    } catch (_: Exception) {
        null
    }
}

private fun downloadFile(url: URI, destPath: Path) {
    Files.createDirectories(destPath.parent)
    var current = url
    repeat(5) {
        val connection = current.toURL().openConnection() as HttpURLConnection
        connection.instanceFollowRedirects = true
        connection.setRequestProperty("User-Agent", "refact-jetbrains")
        val status = connection.responseCode
        if (status in 300..399) {
            val location = connection.getHeaderField("Location")
            if (!location.isNullOrBlank()) {
                current = current.resolve(location)
                return@repeat
            }
        }
        if (status != 200) {
            throw IOException("download failed $status $current")
        }
        connection.inputStream.use { input ->
            Files.copy(input, destPath, StandardCopyOption.REPLACE_EXISTING)
        }
        return
    }
    throw IOException("too many redirects for $url")
}

private fun verifySha256(archivePath: Path, shaPath: Path) {
    val expected = Regex("[a-fA-F0-9]{64}")
        .find(Files.readString(shaPath))
        ?.value
        ?.lowercase()
        ?: throw IOException("sha256 sidecar did not contain a checksum")
    val actual = sha256(archivePath)
    if (actual != expected) {
        throw IOException("sha256 mismatch for ${archivePath.fileName}")
    }
}

private fun sha256(filePath: Path): String {
    val digest = MessageDigest.getInstance("SHA-256")
    Files.newInputStream(filePath).use { input ->
        val buffer = ByteArray(8192)
        while (true) {
            val read = input.read(buffer)
            if (read < 0) break
            digest.update(buffer, 0, read)
        }
    }
    return digest.digest().joinToString("") { "%02x".format(it) }
}

private fun extractArchive(archivePath: Path, destDir: Path, isWindows: Boolean) {
    if (isWindows) {
        extractZip(archivePath, destDir)
        return
    }
    val process = ProcessBuilder("tar", "-xzf", archivePath.toString(), "-C", destDir.toString())
        .redirectErrorStream(true)
        .start()
    val output = process.inputStream.bufferedReader().readText()
    if (!process.waitFor(120, TimeUnit.SECONDS)) {
        process.destroyForcibly()
        throw IOException("tar extraction timed out")
    }
    if (process.exitValue() != 0) {
        throw IOException("tar extraction failed: $output")
    }
}

private fun extractZip(archivePath: Path, destDir: Path) {
    ZipInputStream(Files.newInputStream(archivePath)).use { zip ->
        while (true) {
            val entry = zip.nextEntry ?: break
            val target = destDir.resolve(entry.name).normalize()
            if (!target.startsWith(destDir)) {
                throw IOException("zip entry escapes target directory: ${entry.name}")
            }
            if (entry.isDirectory) {
                Files.createDirectories(target)
            } else {
                Files.createDirectories(target.parent)
                Files.copy(zip, target, StandardCopyOption.REPLACE_EXISTING)
            }
            zip.closeEntry()
        }
    }
}

private fun makeExecutable(binPath: Path) {
    if (!SystemInfo.isWindows) {
        binPath.toFile().setExecutable(true, false)
    }
}

private fun deleteRecursively(root: Path) {
    if (!Files.exists(root)) return
    Files.walk(root).use { paths ->
        paths.sorted(Comparator.reverseOrder()).forEach { path -> Files.deleteIfExists(path) }
    }
}
